In [2]:
# NeuroShield AI — Data Preprocessing

### Stroke Risk Prediction Dataset

#This notebook performs data preprocessing for the NeuroShield AI project.

### Preprocessing Objectives
#- Load and inspect the raw dataset
#- Check data types and data quality
#- Detect duplicate records
#- Handle missing values
#- Remove unnecessary columns
#- Encode categorical variables
#- Validate the processed dataset
#- Save the cleaned dataset for further analysis and model development
    

In [3]:
import pandas as pd
import numpy as np

In [5]:
# Load the raw stroke dataset
df = pd.read_csv("../dataset/raw/stroke.csv")

print("Dataset loaded successfully.")

Dataset loaded successfully.


In [6]:
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

Number of rows: 5110
Number of columns: 12


In [7]:
df.head()


,id,gender,age,hypertension,heart_disease,ever_married,work_type,Residence_type,avg_glucose_level,bmi,smoking_status,stroke
0,9046,Male,67.0,0,1,Yes,Private,Urban,228.69,36.6,formerly smoked,1
1,51676,Female,61.0,0,0,Yes,Self-employed,Rural,202.21,NaN,never smoked,1
2,31112,Male,80.0,0,1,Yes,Private,Rural,105.92,32.5,never smoked,1
3,60182,Female,49.0,0,0,Yes,Private,Urban,171.23,34.4,smokes,1
4,1665,Female,79.0,1,0,Yes,Self-employed,Rural,174.12,24.0,never smoked,1


In [8]:
print("Columns in the dataset:")
print(df.columns.tolist())

Columns in the dataset:
['id', 'gender', 'age', 'hypertension', 'heart_disease', 'ever_married', 'work_type', 'Residence_type', 'avg_glucose_level', 'bmi', 'smoking_status', 'stroke']


In [9]:
# Check the data types of all columns
print("Data types of each column:")
print(df.dtypes)

Data types of each column:
id                     int64
gender                object
age                  float64
hypertension           int64
heart_disease          int64
ever_married          object
work_type             object
Residence_type        object
avg_glucose_level    float64
bmi                  float64
smoking_status        object
stroke                 int64
dtype: object


In [10]:
# Check for missing values
print("Missing values in each column:")
print(df.isnull().sum())

Missing values in each column:
id                     0
gender                 0
age                    0
hypertension           0
heart_disease          0
ever_married           0
work_type              0
Residence_type         0
avg_glucose_level      0
bmi                  201
smoking_status         0
stroke                 0
dtype: int64


In [11]:
# Calculate missing-value percentage
missing_percentage = (df.isnull().sum() / len(df)) * 100

print("Missing-value percentage:")
print(missing_percentage)

Missing-value percentage:
id                   0.000000
gender               0.000000
age                  0.000000
hypertension         0.000000
heart_disease        0.000000
ever_married         0.000000
work_type            0.000000
Residence_type       0.000000
avg_glucose_level    0.000000
bmi                  3.933464
smoking_status       0.000000
stroke               0.000000
dtype: float64


In [12]:
# Check for duplicate records
duplicate_count = df.duplicated().sum()

print("Number of duplicate records:", duplicate_count)

Number of duplicate records: 0


In [13]:
# Generate a compact summary of the dataset
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 5110 non-null   int64  
 1   gender             5110 non-null   object 
 2   age                5110 non-null   float64
 3   hypertension       5110 non-null   int64  
 4   heart_disease      5110 non-null   int64  
 5   ever_married       5110 non-null   object 
 6   work_type          5110 non-null   object 
 7   Residence_type     5110 non-null   object 
 8   avg_glucose_level  5110 non-null   float64
 9   bmi                4909 non-null   float64
 10  smoking_status     5110 non-null   object 
 11  stroke             5110 non-null   int64  
dtypes: float64(3), int64(4), object(5)
memory usage: 479.2+ KB


## 5. Handling Duplicate Records

Duplicate records were checked using complete-row comparison.

No duplicate records were found in the raw dataset. The duplicate-removal operation is still applied to ensure the preprocessing pipeline remains robust and reproducible.

In [14]:
# Remove duplicate records
df = df.drop_duplicates()

print("Dataset shape after duplicate removal:", df.shape)

Dataset shape after duplicate removal: (5110, 12)


## 6. Handling Missing BMI Values

The BMI feature contains missing observations. Instead of removing these records, missing BMI values are imputed using the median BMI calculated from the available observations.

Median imputation was selected because it is less sensitive to extreme values than mean imputation and helps preserve the number of observations available for subsequent analysis and model development.

In [15]:
# Calculate the median BMI using available values
bmi_median = df["bmi"].median()

print("Median BMI:", bmi_median)

Median BMI: 28.1


In [16]:
df = df.drop_duplicates()
print("Dataset shape after duplicate removal:", df.shape)

Dataset shape after duplicate removal: (5110, 12)


In [17]:
bmi_median = df["bmi"].median()
print("Median BMI:", bmi_median)


Median BMI: 28.1


In [18]:
# Fill missing BMI values using the median
df["bmi"] = df["bmi"].fillna(bmi_median)

print("Missing BMI values after imputation:", df["bmi"].isnull().sum())

Missing BMI values after imputation: 0


In [19]:
# Verify dataset shape after BMI imputation
print("Dataset shape after BMI imputation:", df.shape)

Dataset shape after BMI imputation: (5110, 12)


In [20]:
# Verify missing values across the entire dataset
print("Missing values after preprocessing:")
print(df.isnull().sum())

Missing values after preprocessing:
id                   0
gender               0
age                  0
hypertension         0
heart_disease        0
ever_married         0
work_type            0
Residence_type       0
avg_glucose_level    0
bmi                  0
smoking_status       0
stroke               0
dtype: int64


## 7. Removing Unnecessary Columns

The `id` column is a unique identifier and does not represent a meaningful clinical or lifestyle feature for stroke prediction. Therefore, it is removed before further processing.

In [21]:
# Remove the unique identifier
df = df.drop(columns=["id"])

print("Dataset shape after removing id:", df.shape)
print("\nRemaining columns:")
print(df.columns.tolist())

Dataset shape after removing id: (5110, 11)

Remaining columns:
['gender', 'age', 'hypertension', 'heart_disease', 'ever_married', 'work_type', 'Residence_type', 'avg_glucose_level', 'bmi', 'smoking_status', 'stroke']


In [22]:
# Inspect categorical columns and their unique values

categorical_columns = df.select_dtypes(include="object").columns

for column in categorical_columns:
    print(f"\n{column}:")
    print(df[column].value_counts())


gender:
gender
Female    2994
Male      2115
Other        1
Name: count, dtype: int64

ever_married:
ever_married
Yes    3353
No     1757
Name: count, dtype: int64

work_type:
work_type
Private          2925
Self-employed     819
children          687
Govt_job          657
Never_worked       22
Name: count, dtype: int64

Residence_type:
Residence_type
Urban    2596
Rural    2514
Name: count, dtype: int64

smoking_status:
smoking_status
never smoked       1892
Unknown            1544
formerly smoked     885
smokes              789
Name: count, dtype: int64


## 8. Encoding Categorical Variables

Categorical variables are converted into numerical representations so that they can be used by machine learning algorithms.

One-hot encoding is used for categorical features because the categories do not have an inherent numerical order. The target variable `stroke` is retained as a binary numerical variable.

In [23]:
# Encode categorical variables using one-hot encoding

categorical_columns = df.select_dtypes(include="object").columns

df = pd.get_dummies(
    df,
    columns=categorical_columns,
    drop_first=False,
    dtype=int
)

print("Dataset shape after categorical encoding:", df.shape)
print("\nEncoded columns:")
print(df.columns.tolist())

Dataset shape after categorical encoding: (5110, 22)

Encoded columns:
['age', 'hypertension', 'heart_disease', 'avg_glucose_level', 'bmi', 'stroke', 'gender_Female', 'gender_Male', 'gender_Other', 'ever_married_No', 'ever_married_Yes', 'work_type_Govt_job', 'work_type_Never_worked', 'work_type_Private', 'work_type_Self-employed', 'work_type_children', 'Residence_type_Rural', 'Residence_type_Urban', 'smoking_status_Unknown', 'smoking_status_formerly smoked', 'smoking_status_never smoked', 'smoking_status_smokes']


In [24]:
# Final dataset validation

print("Final dataset shape:", df.shape)

print("\nMissing values:")
print(df.isnull().sum().sum())

print("\nDuplicate records:")
print(df.duplicated().sum())

print("\nData types:")
print(df.dtypes)

Final dataset shape: (5110, 22)

Missing values:
0

Duplicate records:
0

Data types:
age                               float64
hypertension                        int64
heart_disease                       int64
avg_glucose_level                 float64
bmi                               float64
stroke                              int64
gender_Female                       int64
gender_Male                         int64
gender_Other                        int64
ever_married_No                     int64
ever_married_Yes                    int64
work_type_Govt_job                  int64
work_type_Never_worked              int64
work_type_Private                   int64
work_type_Self-employed             int64
work_type_children                  int64
Residence_type_Rural                int64
Residence_type_Urban                int64
smoking_status_Unknown              int64
smoking_status_formerly smoked      int64
smoking_status_never smoked         int64
smoking_status_smokes           

In [25]:
# Verify the target variable

print("Stroke target values:")
print(df["stroke"].value_counts())

print("\nTarget data type:", df["stroke"].dtype)

Stroke target values:
stroke
0    4861
1     249
Name: count, dtype: int64

Target data type: int64


In [26]:
# Save the processed dataset

output_path = "../dataset/processed/cleaned_stroke_dataset.csv"

df.to_csv(output_path, index=False)

print("Cleaned dataset saved successfully.")
print("File:", output_path)

Cleaned dataset saved successfully.
File: ../dataset/processed/cleaned_stroke_dataset.csv


In [27]:
# Preview the final processed dataset

df.head()

,age,hypertension,heart_disease,avg_glucose_level,bmi,stroke,gender_Female,gender_Male,gender_Other,ever_married_No,...,work_type_Never_worked,work_type_Private,work_type_Self-employed,work_type_children,Residence_type_Rural,Residence_type_Urban,smoking_status_Unknown,smoking_status_formerly smoked,smoking_status_never smoked,smoking_status_smokes
0,67.0,0,1,228.69,36.6,1,0,1,0,0,...,0,1,0,0,0,1,0,1,0,0
1,61.0,0,0,202.21,28.1,1,1,0,0,0,...,0,0,1,0,1,0,0,0,1,0
2,80.0,0,1,105.92,32.5,1,0,1,0,0,...,0,1,0,0,1,0,0,0,1,0
3,49.0,0,0,171.23,34.4,1,1,0,0,0,...,0,1,0,0,0,1,0,0,0,1
4,79.0,1,0,174.12,24.0,1,1,0,0,0,...,0,0,1,0,1,0,0,0,1,0


## 9. Final Preprocessing Summary

The raw Stroke Prediction Dataset was processed through the following steps:

1. Loaded the raw dataset containing 5,110 records and 12 columns.
2. Checked the dataset structure, data types, and data quality.
3. Checked for duplicate records; no duplicates were found.
4. Identified 201 missing BMI values.
5. Imputed missing BMI values using the median BMI of 28.1.
6. Removed the `id` column because it is a unique identifier and is not used as a predictive feature.
7. Converted categorical variables into numerical features using one-hot encoding.
8. Retained the `Unknown` category in `smoking_status` rather than assigning it another meaning.
9. Validated the processed dataset to ensure there were no remaining missing values or duplicate records.
10. Saved the final processed dataset as `cleaned_stroke_dataset.csv`.

### Final Dataset

- Original dataset: 5,110 rows × 12 columns
- Final dataset: 5,110 rows × 22 columns
- Missing values: 0
- Duplicate records: 0
- Target variable: `stroke`
- Processed file: `dataset/processed/cleaned_stroke_dataset.csv`

The processed dataset is now ready for Exploratory Data Analysis and subsequent machine learning tasks.

In [28]:
# Verify the saved processed dataset

processed_df = pd.read_csv("../dataset/processed/cleaned_stroke_dataset.csv")

print("Saved dataset shape:", processed_df.shape)
print("Missing values:", processed_df.isnull().sum().sum())
print("Duplicate records:", processed_df.duplicated().sum())
print("Target column present:", "stroke" in processed_df.columns)

Saved dataset shape: (5110, 22)
Missing values: 0
Duplicate records: 0
Target column present: True


In [29]:
# Final validation of the processed dataset

expected_columns = [
    "age",
    "hypertension",
    "heart_disease",
    "avg_glucose_level",
    "bmi",
    "stroke",
    "gender_Female",
    "gender_Male",
    "gender_Other",
    "ever_married_No",
    "ever_married_Yes",
    "work_type_Govt_job",
    "work_type_Never_worked",
    "work_type_Private",
    "work_type_Self-employed",
    "work_type_children",
    "Residence_type_Rural",
    "Residence_type_Urban",
    "smoking_status_Unknown",
    "smoking_status_formerly smoked",
    "smoking_status_never smoked",
    "smoking_status_smokes"
]

print("========== FINAL PREPROCESSING CHECK ==========")

print("\n1. Dataset shape:")
print(df.shape)

print("\n2. Number of columns:")
print(len(df.columns))

print("\n3. Column names:")
for i, column in enumerate(df.columns, start=1):
    print(f"{i:2}. {column}")

print("\n4. Missing values:")
print(df.isnull().sum().sum())

print("\n5. Duplicate records:")
print(df.duplicated().sum())

print("\n6. ID column present:")
print("id" in df.columns)

print("\n7. Target column present:")
print("stroke" in df.columns)

print("\n8. Column structure matches expected:")
print(list(df.columns) == expected_columns)

print("\n========== CHECK COMPLETE ==========")

========== FINAL PREPROCESSING CHECK ==========

1. Dataset shape:
(5110, 22)

2. Number of columns:
22

3. Column names:
 1. age
 2. hypertension
 3. heart_disease
 4. avg_glucose_level
 5. bmi
 6. stroke
 7. gender_Female
 8. gender_Male
 9. gender_Other
10. ever_married_No
11. ever_married_Yes
12. work_type_Govt_job
13. work_type_Never_worked
14. work_type_Private
15. work_type_Self-employed
16. work_type_children
17. Residence_type_Rural
18. Residence_type_Urban
19. smoking_status_Unknown
20. smoking_status_formerly smoked
21. smoking_status_never smoked
22. smoking_status_smokes

4. Missing values:
0

5. Duplicate records:
0

6. ID column present:
False

7. Target column present:
True

8. Column structure matches expected:
True

========== CHECK COMPLETE ==========


In [30]:
# Final compact validation

print("========== NEUROSHIELD AI PREPROCESSING ==========")
print(f"Rows: {df.shape[0]}")
print(f"Columns: {df.shape[1]}")
print(f"Missing values: {df.isnull().sum().sum()}")
print(f"Duplicate records: {df.duplicated().sum()}")
print(f"ID column present: {'id' in df.columns}")
print(f"Stroke target present: {'stroke' in df.columns}")
print(f"Processed file exists: {__import__('os').path.exists('../dataset/processed/cleaned_stroke_dataset.csv')}")
print("===================================================")

========== NEUROSHIELD AI PREPROCESSING ==========
Rows: 5110
Columns: 22
Missing values: 0
Duplicate records: 0
ID column present: False
Stroke target present: True
Processed file exists: True
